### ISSUE TITLE
Copy Image does not support transparency

### ISSUE THREAD

**Issue by gentlecolts** [Yellow-Dog-Man/Resonite-Issues#1533]
### Is your feature request related to a problem? Please describe.

When using copy image on an image with transparency, the pasted image does not retain transparent info

### Describe the solution you'd like

copy transparency as well if present

### Describe alternatives you've considered

n/a

### Requesters

zangooseoo

**Comment by Frooxius** [comment]
What software are you using this with?

We actually do copy the transparency data into the clipboard through the API's, but transparency doesn't seem to be very well supported through the clipboard with most software.

E.g. you can see this when you copy image with transparency from Photoshop and paste it into MS Paint (the new versions support alpha) or Resonite.

Does the software you use support copying transparency properly?

**Comment by gentlecolts** [comment]
Sorry if I was not clear, I was referring to the radial menu "Copy Image" button ingame

**Comment by Frooxius** [comment]
I understand that part. I am asking which software are you copying the images into? Are you pasting them to Photoshop? GIMP? Something else?

**Comment by jae1911** [comment]
I can reproduce that one by copying for instance `resdb:///bcf2492766d9153b4e812f4079d4233bbc3ba521ea51dbffa547271f71ee9b73.webp` (my logo) to Discord, which goes from:
![Resonite_4e6YSPJYtF](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/10509376-4a91-4f07-9ce9-f71049c5fe46)

To:
![Discord_ECMmtoypU9](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/07422b01-1512-451f-8fe9-2716410702f5)

While it should look like:
![Discord_mDVEfGI1tn](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/14e4035c-ebd6-4091-a2ea-fc19ba16317c)

**Comment by Frooxius** [comment]
@jae1911 Can you successfully copy the image data with transparency into Discord from another software (e.g. Photoshop or GIMP) and have it retain its transparency? 

Please read my original posts in full, because this is something I asked for. It could save a lot of this back and forth.

**Comment by jae1911** [comment]
@Frooxius yes, I can.

Tested with Affinity Photo.

Original:
![Photo_9XbmcLFvp3](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/09c5406c-1c99-4d44-9fcc-e208d6c7d9cb)

Discord (had to use light theme for that one):
![Discord_WdClPytoCb](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/b12f4dc3-ad29-4ebd-a64f-eef6afebdb6b)

**Comment by gentlecolts** [comment]
> which software are you copying the images into

The particular image that prompted this was copying from resonite into resonite.  It was for an image in my inventory who's link had died, but I still have it cached, so I was hoping to save a raw version of it to my inventory

**Comment by Frooxius** [comment]
@jae1911 I see thanks! Does it work the other way? Transparent image from Resonite to Affinity Photo?

I'm basing this mostly on the fact that when I try this with Photoshop, it doesn't work (e.g. transparent image from Photoshop to Discord loses the transparency).

@gentlecolts Hmm, that might be something there if the roundtrip doesn't survive the data, we can check that.

You could try exporting the image though through the file browser, that can potentially give you the file in this particular case.

**Comment by jae1911** [comment]
> Transparent image from Resonite to Affinity Photo?

Alright, good discovery.
By using File -> New From Clipboard, transparency is retained:
![Photo_AQZpArNCza](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/30d7d5c6-8ce9-43a5-9b1f-38e777f6685b)

**However**, it seems when using a web-based platform, for instance, when pasting into Discord (based on Chromium), a background is added and same thing can be reproduced in Firefox (for instance):
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/86b347c3-a5bb-49fd-bc48-5cffd514871a)

**Comment by Frooxius** [comment]
@jae1911 Ah that's good info, thank you!

This sounds a bit like it's something on the receiving end in part. I know there are multiple formats for the clipboard image data, perhaps some are more compatible than others?

I could poke around this some and see if different formats have different effects.

**Comment by jae1911** [comment]
> This sounds a bit like it's something on the receiving end in part

Yeah, in the end it seems up to the receiving software implementation.

As a last test, I tested with Krita, which also adds the background as well:
![krita_ciiFdnFGnf](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/76598503/02deff43-37e7-450f-b62f-ad145a8bcc4b)

Either Affinity Photo is doing some kind of magic or all other software is missing something.

All the tests were of course conducted with the same image: `resdb:///bcf2492766d9153b4e812f4079d4233bbc3ba521ea51dbffa547271f71ee9b73.webp`

**Comment by PJB3005** [comment]
Resonite uses Winforms' `Clipboard.SetImage()` to set the bitmap contents, and this is the clipboard formats it produces:

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8107459/a6cbadc0-4888-4fca-a304-8fbfc367e924)

[`CF_BITMAP`](https://learn.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats) is some ancient Windows type that I don't think supports transparency, and I assume the grey background color we're seeing gets baked in by Winforms. `System.Drawing.Bitmap` appears to be some custom format that wraps a PNG and *does* support transparency (when checking with ShareX) but probably isn't used by non-winforms apps.

For comparison, here's Discord, Krita and Firefox respectively:

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8107459/9ffb8d0e-d591-4f67-b97a-a75256a56539)

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8107459/6a63931f-94db-4024-b635-79c117ba0ed2)

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8107459/43591155-e766-423f-916b-4c60fa53f3c7)

All combinations of copy pasting seem to work fine between Krita, Discord and MS Paint (Windows 11), but Firefox seems really buggy. Firefox -> Discord works, Firefox -> Paint turns alpha black, and Firefox -> Krita has the same if I use the "paste attached bitmap" option instead of "paste original location" (firefox specifies original image URL in clipboard data). 

Anyways, the common trend here appears to be `CF_DIB` and `CF_DIBV5` as "formats that need supporting".

**Comment by Frooxius** [comment]
Thanks for investigating this! This additional info helps a lot!

**Comment by lxw404** [comment]
I just experienced this using the `Copy Image` radial option on a newly imported image, then pasting directly back into Resonite, and it not only discards transparency, but also the background color is completely wrong:

![Resonite_rbsGIf8Cw8](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/52231149/35542eb0-05f6-4fac-89bc-8caa05ed44c2)

Here is the original image:
![BasicCircle001](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/52231149/cb96290b-dea2-4c7c-88f4-d3b8123d9bf7)

Also a log of the action:
[Pasted Alpha Image From Resonite To Resonite - 2024.5.3.561 - 2024-05-03 06_52_28.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/15199061/Pasted.Alpha.Image.From.Resonite.To.Resonite.-.2024.5.3.561.-.2024-05-03.06_52_28.log)

**Comment by Frooxius** [comment]
Hmm I'm looking into this some more, but it looks it's a fair bit more tricky.

We don't really have control over how exactly the data is being set. We're writing the alpha data and we're using `PixelFormat.Format32bppArgb` for the data we're sending to the clipboard.

I don't see anything that would immediately allow a different format or to set the formats as you show. We might need to convert the clipboard helper to a .NET 8 process perhaps and use more modern API or find if there's alternate NuGet package for setting clipboard data that supports these newer formats.
